libdnf/sack/query.cpp

```diff
@@ -0,0 +1,1821 @@
+            break;
+
+        s_installed = pool_id2solvable(pool, id_installed);
+        matched = false;
+        Id id_available = -1;
+        while (true) {
+            id_available = resultAvailable->next(id_available);
+            if (id_available == -1)
+                break;
+
+            s_available = pool_id2solvable(pool, id_available);
+            if ((s_installed->name == s_available->name) && (s_installed->evr == s_available->evr)
+                && (s_installed->arch == s_available->arch)) {
+                matched = true;
+                break;
+            }
+        }
+        if (!matched)
+            MAPSET(resultMap, id_installed);
+    }
+}
+
+void
+Query::filterRecent(const long unsigned int recent_limit)
+{
+    apply();
+    auto resultPset = pImpl->result.get();
+    auto resultMap = pImpl->result->getMap();
+
+    Id id = -1;
+    while (true) {
+        id = resultPset->next(id);
+        if (id == -1)
+                break;
+        DnfPackage *pkg = dnf_package_new(pImpl->sack, id);
+        guint64 build_time = dnf_package_get_buildtime(pkg);
+        if (build_time <= recent_limit) {
+            MAPCLR(resultMap, id);
+        }
+    }
+}
+
+void
+Query::filterDuplicated()
+{
+    Queue samename;
+    Pool *pool = dnf_sack_get_pool(pImpl->sack);
+
+    addFilter(HY_PKG_REPONAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+    apply();
+
+    auto resultMap = pImpl->result->getMap();
+    hy_query_to_name_ordered_queue(this, &samename);
+
+    Solvable *considered, *highest = 0;
+    int start_block = -1;
+    int i;
+    MAPZERO(resultMap);
+    for (i = 0; i < samename.count; ++i) {
+        Id p = samename.elements[i];
+        considered = pool->solvables + p;
+        if (!highest || highest->name != considered->name) {
+            /* start of a new block */
+            if (start_block == -1) {
+                highest = considered;
+                start_block = i;
+                continue;
+            }
+            if (start_block != i - 1) {
+                add_duplicates_to_map(pool, resultMap, samename, start_block,
+                                      i);
+            }
+            highest = considered;
+            start_block = i;
+        }
+    }
+    if (start_block != -1) {
+        add_duplicates_to_map(pool, resultMap, samename, start_block, i);
+    }
+}
+
+}
+
+void
+hy_query_to_name_ordered_queue(HyQuery query, Queue *samename)
+{
+    hy_query_apply(query);
+    Pool *pool = dnf_sack_get_pool(query->getSack());
+
+    queue_init(samename);
+    for (int i = 1; i < pool->nsolvables; ++i)
+        if (MAPTST(query->getResult(), i))
+            queue_push(samename, i);
+
+    solv_sort(samename->elements, samename->count, sizeof(Id), libdnf::filter_latest_sortcmp, pool);
+}
+
+void
+hy_query_to_name_arch_ordered_queue(HyQuery query, Queue *samename)
+{
+    hy_query_apply(query);
+    Pool *pool = dnf_sack_get_pool(query->getSack());
+
+    queue_init(samename);
```

Comment:
Optimise: call query->getResult() outside for loop.
const auto * result = query->getResult();